python/paddle/distributed/auto_parallel/api.py

```diff
@@ -223,3 +234,156 @@ def reshard(dist_tensor, dist_attr):
         raise RuntimeError(
             "paddle.dist.reshard only support dynamic graph now. It will be supported for static graph later."
         )
+
+
+def shard_layer(
+    layer: nn.Layer,
+    process_mesh: dist.ProcessMesh,
+    shard_fn: Callable = None,
+    input_fn: Callable = None,
+    output_fn: Callable = None,
+) -> nn.Layer:
+    """
+    Converts all layer's parameters to DistTensor parameters according to
+    the `shard_fn` specified. It could also control the conversion of input
+    or output of the layer by specifying the `input_fn` and `output_fn`.
+    (i.e. convert the input to `paddle.Tensor` with DistTensor, convert output
+    back to `paddle.Tensor` with DenseTensor.)
+
+    The `shard_fn` should have the following signature:
+
+        def shard_fn(layer_name, layer, process_mesh) -> None
+
+    The `input_fn` should have the following signature:
+
+        def input_fn(inputs, process_mesh) -> list(paddle.Tensor)
+
+    In general, the type of `input_fn` return value is paddle.Tensor with DistTensor.
+
+    The `output_fn` should have the following signature:
+
+        def output_fn(outputs, process_mesh) -> list(paddle.Tensor)
+
+    In general, the type of `output_fn` return value is paddle.Tensor with DenseTensor.
+
+    Args:
+        layer (paddle.nn.Layer): The Layer object to be shard.
+        process_mesh (paddle.distributed.ProcessMesh): The `ProcessMesh` information
+            to be place the input `layer`.
+        shard_fn (Callable): The function to shard layer parameters across
+            the `process_mesh`. If not specified, by default we replicate
+            all parameters of the layer across the `process_mesh`.
+        input_fn (Callable): Specify how the input of the layer is sharded.
+            The `input_fn` will be registered for the Layer as a `forward pre-hook`.
+            By default we do not shard the input.
+        output_fn (Callable): Specify how the output of the layer is sharded or
+            convert it back to `paddle.Tensor` with DenseTensor.
+            The `output_fn` will be registered for the Layer as `forward post-hook`.
+            By default we do not shard or convert the output.
+    Returns:
+        Layer: A layer that contains parameters/buffers
+            that are all `paddle.Tensor` with DistTensor
+
+    Examples:
+        ..code-block:: python
```

Comment:
```suggestion
        .. code-block:: python
```
<img width="931" alt="image" src="https://github.com/PaddlePaddle/Paddle/assets/70642955/426ae996-7351-43a9-86a3-33c2554019b1">